File: src/packages/agent_core/builtin_tools/web_fetch/jina.py
from __future__ import annotations

import httpx

from .provider import WebFetchResult

_DEFAULT_JINA_BASE_URL = "https://r.jina.ai"


class JinaWebFetchHttpError(RuntimeError):
    def __init__(self, *, status_code: int, message: str) -> None:
        super().__init__(message)
        self.status_code = status_code


class JinaWebFetchProvider:
    def __init__(
        self,
        *,
        api_key: str,
        base_url: str = _DEFAULT_JINA_BASE_URL,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        cleaned = api_key.strip()
        if not cleaned:
            raise ValueError("Jina api_key 不能为空")
        self._api_key = cleaned
        self._base_url = base_url.rstrip("/")
        self._client = client

    async def fetch(self, *, url: str, max_length: int) -> WebFetchResult:
        client = self._client
        if client is None:
            timeout = httpx.Timeout(None)
            async with httpx.AsyncClient(timeout=timeout) as client:
                return await self._fetch_with_client(client=client, url=url, max_length=max_length)
        return await self._fetch_with_client(client=client, url=url, max_length=max_length)

    async def _fetch_with_client(
        self,
        *,
        client: httpx.AsyncClient,
        url: str,
        max_length: int,
    ) -> WebFetchResult:
        headers: dict[str, str] = {}
        if self._api_key:
            headers["Authorization"] = f"Bearer {self._api_key}"

        request_url = f"{self._base_url}/{url}"
        resp = await client.get(request_url, headers=headers)
        if resp.status_code < 200 or resp.status_code >= 300:
            raise JinaWebFetchHttpError(
                status_code=int(resp.status_code),
                message=f"Jina Reader 返回 {resp.status_code}",
            )

        content = (resp.text or "").strip()
        title = _extract_title_from_markdown(content)

        truncated = False
        if len(content) > max_length:
            content = content[:max_length]
            truncated = True
        if len(title) > 512:
            title = title[:512]

        return WebFetchResult(url=url, content=content, title=title, truncated=truncated)


def _extract_title_from_markdown(text: str) -> str:
    for line in text.splitlines():
        stripped = line.strip()
        if not stripped:
            continue
        if stripped.startswith("# "):
            return stripped[2:].strip()
        lowered = stripped.casefold()
        if lowered.startswith("title:"):
            return stripped.split(":", 1)[1].strip()
        return ""
    return ""


__all__ = [
    "JinaWebFetchHttpError",
    "JinaWebFetchProvider",
]

File: src/services/worker/internal/tools/builtin/web_fetch/jina.go
package webfetch

import (
	"context"
	"io"
	"net/http"
	"strings"
	"time"
)

const defaultJinaBaseURL = "https://r.jina.ai"

type JinaProvider struct {
	apiKey  string
	baseURL string
	client  *http.Client
}

// NewJinaProvider creates a Jina provider. apiKey may be empty for anonymous
// free-tier access (rate-limited but requires no account).
func NewJinaProvider(apiKey string) (*JinaProvider, error) {
	return &JinaProvider{
		apiKey:  strings.TrimSpace(apiKey),
		baseURL: defaultJinaBaseURL,
		client:  &http.Client{Timeout: 60 * time.Second},
	}, nil
}

func (p *JinaProvider) Fetch(ctx context.Context, targetURL string, maxLength int) (Result, error) {
	cleanedURL := strings.TrimSpace(targetURL)
	reqURL := strings.TrimRight(p.baseURL, "/") + "/" + cleanedURL

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL, nil)
	if err != nil {
		return Result{}, err
	}
	if p.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+p.apiKey)
	}

	resp, err := p.client.Do(req)
	if err != nil {
		return Result{}, err
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return Result{}, HttpError{StatusCode: resp.StatusCode}
	}

	body, err := io.ReadAll(io.LimitReader(resp.Body, 5_000_000))
	if err != nil {
		return Result{}, err
	}

	content := strings.TrimSpace(string(body))
	title := extractTitleFromMarkdown(content)

	truncated := false
	if len(content) > maxLength {
		content = content[:maxLength]
		truncated = true
	}
	if len(title) > 512 {
		title = title[:512]
	}

	return Result{
		URL:       cleanedURL,
		Title:     title,
		Content:   content,
		Truncated: truncated,
	}, nil
}

func extractTitleFromMarkdown(text string) string {
	for _, line := range strings.Split(text, "\n") {
		stripped := strings.TrimSpace(line)
		if stripped == "" {
			continue
		}
		if strings.HasPrefix(stripped, "# ") {
			return strings.TrimSpace(stripped[2:])
		}
		lowered := strings.ToLower(stripped)
		if strings.HasPrefix(lowered, "title:") {
			return strings.TrimSpace(strings.SplitN(stripped, ":", 2)[1])
		}
		return ""
	}
	return ""
}
